src/Compilers/Core/Portable/SourceGeneration/AdditionalSourcesCollection.cs

```diff
@@ -70,6 +84,9 @@ public void Add(string hintName, SourceText source)
                 throw new ArgumentException(string.Format(CodeAnalysisResources.SourceTextRequiresEncoding, hintName), nameof(source));
             }
 
+            Debug.Assert(Path.GetFullPath(hintName).StartsWith(Environment.CurrentDirectory, _hintNameComparison),
```

Comment:
It seems reasonable to me to do this check at runtime. But I am also wondering if there is some sane way to check that the `hintName` is "contained in" the CurrentDirectory in some sense. File systems are weird and different across platforms and so on..so I could imagine someone throwing a hard link into the current directory, or adding some separator that we happen to allow but which actually means to redirect to somewhere, and then the user's files "escaping" the expected root.

Perhaps there's a point where we say, ok, you were obviously trying to circumvent this, congrats, you've hurt yourself.

Reply:
> It seems reasonable to me to do this check at runtime.

You mean adding an `if` and throwing an exception? This assert resolves the `hintName` against workdir. But in reality, `hintName` will be resolved against `$(CompilerGeneratedFilesOutputPath)`. So I'm wondering if the check would be wrong in some edge cases.

> wondering if there is some sane way to check that the `hintName` is "contained in" the CurrentDirectory

You mean better than `Path.GetFullPath(hintName).StartsWith(Environment.CurrentDirectory, _hintNameComparison)`? I don't know, the runtime doesn't provide many `Path` APIs, so I think some hack like this is necessary.

In general, I don't think the goal here is to catch all cases. If the user wants to escape current directory, I guess he can use reflection to force-add hintName like `../file.cs`. So the goal is to merely catch common cases. And until the runtime provides better `Path` APIs for this (tested against many different file systems), this is the best we can do, I think.